# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources in the README.md file
- Happy learning!

In [1]:
# Import your libraries:

%matplotlib inline

import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
from sqlalchemy import create_engine



# Challenge 0 - Load,Query and Create connection of your Dataset

#### A lot of the times you won't have files already saved in an Excel or CSV file for you to prepare your data, implying that a lot of times you'll be extracting data from a Data Warehouse or Data Lake, majority of times through SQL.
#### A couple of times you may want to do some queries on your table in mySQL to get a slight view on what you have in hands,
#### so let's simulate that!

#### First we'll need to create a database and table in mySQL:

##### 1º- Open the austin_weather.sql file in MySQL Workbench and run the script into a desired schema.

#### 2º- As we are in mySQL Workbench, we can do some queries there, to get an overview on some characteristics of our data:
 - a) How many days are recorded in the dataset?
 - b) What is the day with the Highest Temperature in Fahrenheit (column TempHighF)
 - c) What is the average Humidity across all days? (column HumidityAvgPercent)
 - d) Top 10 days, where SeaLevelPressureAvgInches is the highest, knowing DewPointAvgF is higher than 28 ?

In [2]:
# Write your answers below:

# a) 1319
# b) '2017-07-29'
# c) 66.57 Farenheit
# d) 30.54

#### Now that you explored a couple of elements in your table, let's bring your table into this jupyter notebook, by creating a Python-SQL connection like you did on MySQL Project!
#### In case you need a little refresher check this [link](https://www.dataquest.io/blog/sql-insert-tutorial/).
##### 1º - Create a connection using sqlalchemy from python to mysql 
##### 2º- Load the table into a variable called weather_df


In [3]:
# Your code here
import pandas as pd
from sqlalchemy import create_engine
import getpass

# Define the database connection parameters 
password = getpass.getpass('Enter your MySQL password: ')

# Create SQLAlchemy engine
engine = create_engine(f'mysql+pymysql://root:{password}@localhost/weather_data')

# Now use pandas with the engine
weather_df = pd.read_sql_query("SELECT * FROM austin_weather", engine)

print(weather_df.head())


        Date  TempHighF  TempAvgF  TempLowF DewPointHighF DewPointAvgF  \
0 2013-12-21         74        60        45            67           49   
1 2013-12-22         56        48        39            43           36   
2 2013-12-23         58        45        32            31           27   
3 2013-12-24         61        46        31            36           28   
4 2013-12-25         58        50        41            44           40   

  DewPointLowF HumidityHighPercent HumidityAvgPercent HumidityLowPercent  ...  \
0           43                  93                 75                 57  ...   
1           28                  93                 68                 43  ...   
2           23                  76                 52                 27  ...   
3           21                  89                 56                 22  ...   
4           36                  86                 71                 56  ...   

  SeaLevelPressureAvgInches SeaLevelPressureLowInches VisibilityHigh


# Challenge 1 - Describe the Dataset

#### Describe the dataset you have loaded: 
- Look at the variables and their types
- Examine the descriptive statistics of the numeric variables 
- Look at the first five rows of all variables to evaluate the categorical variables as well

In [4]:
weather_df.shape

(1319, 21)

In [5]:
# Your code here
weather_df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1319 entries, 0 to 1318
Data columns (total 21 columns):
 #   Column                      Non-Null Count  Dtype         
---  ------                      --------------  -----         
 0   Date                        1319 non-null   datetime64[ns]
 1   TempHighF                   1319 non-null   int64         
 2   TempAvgF                    1319 non-null   int64         
 3   TempLowF                    1319 non-null   int64         
 4   DewPointHighF               1319 non-null   object        
 5   DewPointAvgF                1319 non-null   object        
 6   DewPointLowF                1319 non-null   object        
 7   HumidityHighPercent         1319 non-null   object        
 8   HumidityAvgPercent          1319 non-null   object        
 9   HumidityLowPercent          1319 non-null   object        
 10  SeaLevelPressureHighInches  1319 non-null   object        
 11  SeaLevelPressureAvgInches   1319 non-null   object      

In [6]:
# Your code here
weather_df.describe()

Date    TempHighF     TempAvgF     TempLowF
count                 1319  1319.000000  1319.000000  1319.000000
mean   2015-10-11 00:00:00    80.862775    70.642911    59.902957
min    2013-12-21 00:00:00    32.000000    29.000000    19.000000
25%    2014-11-15 12:00:00    72.000000    62.000000    49.000000
50%    2015-10-11 00:00:00    83.000000    73.000000    63.000000
75%    2016-09-04 12:00:00    92.000000    83.000000    73.000000
max    2017-07-31 00:00:00   107.000000    93.000000    81.000000
std                    NaN    14.766523    14.045904    14.190648

In [7]:
# Your code here
weather_df.head(5)

Date  TempHighF  TempAvgF  TempLowF DewPointHighF DewPointAvgF  \
0 2013-12-21         74        60        45            67           49   
1 2013-12-22         56        48        39            43           36   
2 2013-12-23         58        45        32            31           27   
3 2013-12-24         61        46        31            36           28   
4 2013-12-25         58        50        41            44           40   

  DewPointLowF HumidityHighPercent HumidityAvgPercent HumidityLowPercent  ...  \
0           43                  93                 75                 57  ...   
1           28                  93                 68                 43  ...   
2           23                  76                 52                 27  ...   
3           21                  89                 56                 22  ...   
4           36                  86                 71                 56  ...   

  SeaLevelPressureAvgInches SeaLevelPressureLowInches VisibilityHighMiles  \
0                     29.68                     29.59                  10   
1                     30.13                     29.87                  10   
2                     30.49                     30.41                  10   
3                     30.45                      30.3                  10   
4                     30.33                     30.27                  10   

  VisibilityAvgMiles VisibilityLowMiles WindHighMPH WindAvgMPH WindGustMPH  \
0                  7                  2          20          4          31   
1                 10                  5          16          6          25   
2                 10                 10           8          3          12   
3                 10                  7          12          4          20   
4                 10                  7          10          2          16   

  PrecipitationSumInches               Events  
0                   0.46  Rain , Thunderstorm  
1                      0                       
2                      0                       
3                      0                       
4                      T                       

[5 rows x 21 columns]

#### Given the information you have learned from examining the dataset, write down three insights about the data in a markdown cell below

#### Your Insights:

1. There are 21 variables in the dataset. 3 of them are numeric and the rest contain some text.

2. The average temperature in Austin ranged between around 70 degrees F and around 93 degrees F. The highest temperature observed during this period was 107 degrees F and the lowest was 19 degrees F.

3. When we look at the head function, we see that a lot of variables contain numeric data even though these columns are of object type. This means we might have to do some data cleansing.


#### Let's examine the DewPointAvgF variable by using the `unique()` function to list all unique values in this dataframe.

Describe what you find in a markdown cell below the code. What did you notice? What do you think made Pandas to treat this column as *object* instead of *int64*? 

In [8]:
# Your code here
weather_df['DewPointAvgF'].unique()


array(['49', '36', '27', '28', '40', '39', '41', '26', '42', '22', '48',
       '32', '8', '11', '45', '55', '61', '37', '47', '25', '23', '20',
       '33', '30', '29', '17', '14', '13', '54', '59', '15', '24', '34',
       '35', '57', '50', '53', '60', '46', '56', '51', '31', '38', '62',
       '43', '63', '64', '67', '66', '58', '70', '68', '65', '69', '71',
       '72', '-', '73', '74', '21', '44', '52', '12', '75', '76', '18'],
      dtype=object)

In [9]:
# Your observation here
# All values are strings: Every value is enclosed in quotes ('49', '36', '27', etc.)
# They're numeric strings: All values appear to be valid numbers, but stored as text
# No obvious missing values: I don't see any 'N/A', '-', or empty strings
# Data type is object: The output shows dtype=object


The following is a list of columns misrepresented as `object`. Use this list to convert the columns to numeric using the `pandas.to_numeric` function in the next cell. If you encounter errors in converting strings to numeric values, you need to catch those errors and force the conversion by supplying `errors='coerce'` as an argument for `pandas.to_numeric`. Coercing will replace non-convertable elements with `NaN` which represents an undefined numeric value. This makes it possible for us to conveniently handle missing values in subsequent data processing.

*Hint: you may use a loop to change one column at a time but it is more efficient to use `apply`.*

In [10]:
wrong_type_columns = ['DewPointHighF', 'DewPointAvgF', 'DewPointLowF', 'HumidityHighPercent', 
                      'HumidityAvgPercent', 'HumidityLowPercent', 'SeaLevelPressureHighInches', 
                      'SeaLevelPressureAvgInches' ,'SeaLevelPressureLowInches', 'VisibilityHighMiles',
                      'VisibilityAvgMiles', 'VisibilityLowMiles', 'WindHighMPH', 'WindAvgMPH', 
                      'WindGustMPH', 'PrecipitationSumInches']

In [11]:
# Your code here
# Apply pd.to_numeric to all columns at once
weather_df[wrong_type_columns] = weather_df[wrong_type_columns].apply(pd.to_numeric, errors='coerce')



#### Check if your code has worked by printing the data types again. You should see only two `object` columns (`Date` and `Events`) now. All other columns should be `int64` or `float64`.

In [12]:
# Your code here
# Check data types after conversion
print("Data types after conversion:")
print(weather_df[wrong_type_columns].dtypes)

# Check for any NaN values that resulted from coercion
print("\nNaN counts per column:")
print(weather_df[wrong_type_columns].isnull().sum())

Data types after conversion:
DewPointHighF                 float64
DewPointAvgF                  float64
DewPointLowF                  float64
HumidityHighPercent           float64
HumidityAvgPercent            float64
HumidityLowPercent            float64
SeaLevelPressureHighInches    float64
SeaLevelPressureAvgInches     float64
SeaLevelPressureLowInches     float64
VisibilityHighMiles           float64
VisibilityAvgMiles            float64
VisibilityLowMiles            float64
WindHighMPH                   float64
WindAvgMPH                    float64
WindGustMPH                   float64
PrecipitationSumInches        float64
dtype: object

NaN counts per column:
DewPointHighF                   7
DewPointAvgF                    7
DewPointLowF                    7
HumidityHighPercent             2
HumidityAvgPercent              2
HumidityLowPercent              2
SeaLevelPressureHighInches      3
SeaLevelPressureAvgInches       3
SeaLevelPressureLowInches       3
VisibilityHighMiles

# Challenge 2 - Handle the Missing Data

#### Now that we have fixed the type mismatch, let's address the missing data.

By coercing the columns to numeric, we have created `NaN` for each cell containing characters. We should choose a strategy to address these missing data.

The first step is to examine how many rows contain missing data.

We check how much missing data we have by applying the `.isnull()` function to our dataset. To find the rows with missing data in any of its cells, we apply `.any(axis=1)` to the function. `austin.isnull().any(axis=1)` will return a column containing true if the row contains at least one missing value and false otherwise. Therefore we must subset our dataframe with this column. This will give us all rows with at least one missing value. 

#### In the next cell, identify all rows containing at least one missing value. Assign the dataframes with missing values to a variable called `missing_values`.

In [13]:
# Your code here
weather_df[wrong_type_columns].isnull()

DewPointHighF  DewPointAvgF  DewPointLowF  HumidityHighPercent  \
0             False         False         False                False   
1             False         False         False                False   
2             False         False         False                False   
3             False         False         False                False   
4             False         False         False                False   
...             ...           ...           ...                  ...   
1314          False         False         False                False   
1315          False         False         False                False   
1316          False         False         False                False   
1317          False         False         False                False   
1318          False         False         False                False   

      HumidityAvgPercent  HumidityLowPercent  SeaLevelPressureHighInches  \
0                  False               False                       False   
1                  False               False                       False   
2                  False               False                       False   
3                  False               False                       False   
4                  False               False                       False   
...                  ...                 ...                         ...   
1314               False               False                       False   
1315               False               False                       False   
1316               False               False                       False   
1317               False               False                       False   
1318               False               False                       False   

      SeaLevelPressureAvgInches  SeaLevelPressureLowInches  \
0                         False                      False   
1                         False                      False   
2                         False                      False   
3                         False                      False   
4                         False                      False   
...                         ...                        ...   
1314                      False                      False   
1315                      False                      False   
1316                      False                      False   
1317                      False                      False   
1318                      False                      False   

      VisibilityHighMiles  VisibilityAvgMiles  VisibilityLowMiles  \
0                   False               False               False   
1                   False               False               False   
2                   False               False               False   
3                   False               False               False   
4                   False               False               False   
...                   ...                 ...                 ...   
1314                False               False               False   
1315                False               False               False   
1316                False               False               False   
1317                False               False               False   
1318                False               False               False   

      WindHighMPH  WindAvgMPH  WindGustMPH  PrecipitationSumInches  
0           False       False        False                   False  
1           False       False        False                   False  
2           False       False        False                   False  
3           False       False        False                   False  
4           False       False        False                    True  
...           ...         ...          ...                     ...  
1314        False       False        False                   False  
1315        False       False        False                   False  
1316        False       False    

In [29]:
# Identify all rows containing at least one missing value
missing_values = weather_df[weather_df.isnull().any(axis=1)]

In [30]:
# Count the number of rows
print(f"Total rows in weather_df: {len(weather_df)}")
print(f"Rows with missing values: {len(missing_values)}")

Total rows in weather_df: 1319
Rows with missing values: 136


In [31]:
# Find the number of missing rows in each column
missing_per_column = weather_df.isna().sum()
print("Missing values per column:")
print(missing_per_column[missing_per_column > 0])  # Only show columns with missing values

Missing values per column:
DewPointHighF                   7
DewPointAvgF                    7
DewPointLowF                    7
HumidityHighPercent             2
HumidityAvgPercent              2
HumidityLowPercent              2
SeaLevelPressureHighInches      3
SeaLevelPressureAvgInches       3
SeaLevelPressureLowInches       3
VisibilityHighMiles            12
VisibilityAvgMiles             12
VisibilityLowMiles             12
WindHighMPH                     2
WindAvgMPH                      2
WindGustMPH                     4
PrecipitationSumInches        124
dtype: int64


There are multiple strategies to handle missing data. Below lists the most common ones data scientists use:

* Removing all rows or all columns containing missing data. This is the simplest strategy. It may work in some cases but not others.

* Filling all missing values with a placeholder value. 
    * For categorical data, `0`, `-1`, and `9999` are some commonly used placeholder values. 
    * For continuous data, some may opt to fill all missing data with the mean. This strategy is not optimal since it can increase the fit of the model.

* Filling the values using some algorithm. 

#### In our case, we will use a hybrid approach which is to first remove the data that contain most missing values then fill in the rest of the missing values with the *linear interpolation* algorithm.

#### Next, count the number of rows of `austin` and `missing_values`.

In [32]:
# Your code here
# Count the number of rows
print(f"Total rows in weather_df: {len(weather_df)}")
print(f"Rows with missing values: {len(missing_values)}")

Total rows in weather_df: 1319
Rows with missing values: 136


#### Calculate the ratio of missing rows to total rows

In [36]:
# Your code here
# Calculate the ratio of missing rows to total rows
missing_ratio = len(missing_values) / len(weather_df)
print(f"Ratio of missing rows to total rows: {missing_ratio:.4f} ({missing_ratio*100:.2f}%)")

Ratio of missing rows to total rows: 0.1031 (10.31%)


As you can see, there is a large proportion of missing data (over 10%). Perhaps we should evaluate which columns have the most missing data and remove those columns. For the remaining columns, we will perform a linear approximation of the missing data.

We can find the number of missing rows in each column using the `.isna()` function. We then chain the `.sum` function to the `.isna()` function and find the number of missing rows per column

In [37]:
# Your code here
# Find the number of missing rows in each column
missing_per_column = weather_df.isna().sum()
print("Missing values per column:")
print(missing_per_column[missing_per_column > 0])  # Only show columns with missing values

Missing values per column:
DewPointHighF                   7
DewPointAvgF                    7
DewPointLowF                    7
HumidityHighPercent             2
HumidityAvgPercent              2
HumidityLowPercent              2
SeaLevelPressureHighInches      3
SeaLevelPressureAvgInches       3
SeaLevelPressureLowInches       3
VisibilityHighMiles            12
VisibilityAvgMiles             12
VisibilityLowMiles             12
WindHighMPH                     2
WindAvgMPH                      2
WindGustMPH                     4
PrecipitationSumInches        124
dtype: int64


#### As you can see from the output, the majority of missing data is in one column called `PrecipitationSumInches`. What's the number of missing values in this column in ratio to its total number of rows?

In [38]:
# Your code here
# Calculate missing ratio for PrecipitationSumInches
precip_missing = weather_df['PrecipitationSumInches'].isna().sum()
precip_total = len(weather_df)
precip_ratio = precip_missing / precip_total

print(f"Missing values in PrecipitationSumInches: {precip_missing}")
print(f"Missing ratio: {precip_ratio:.4f} ({precip_ratio*100:.2f}%)")

Missing values in PrecipitationSumInches: 124
Missing ratio: 0.0940 (9.40%)


Almost 10% data missing! Therefore, we prefer to remove this column instead of filling its missing values. 

#### Remove this column from `austin` using the `.drop()` function. Use the `inplace=True` argument.

*Hints:*

* By supplying `inplace=True` to `drop()`, the original dataframe object will be changed in place and the function will return `None`. In contrast, if you don't supply `inplace=True`, which is equivalent to supplying `inplace=False` because `False` is the default value, the original dataframe object will be kept and the function returns a copy of the transformed dataframe object. In the latter case, you'll have to assign the returned object back to your variable.

* Also, since you are dropping a column instead of a row, you'll need to supply `axis=1` to `drop()`.

[Reference for `pandas.DataFrame.drop`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.drop.html)

In [39]:
# Your code here 
# Remove PrecipitationSumInches column
weather_df.drop('PrecipitationSumInches', axis=1, inplace=True)

# Print `austin` to confirm the column is indeed removed
print("Columns after dropping PrecipitationSumInches:")
print(weather_df.columns.tolist())
print(f"Shape: {weather_df.shape}")


Columns after dropping PrecipitationSumInches:
['Date', 'TempHighF', 'TempAvgF', 'TempLowF', 'DewPointHighF', 'DewPointAvgF', 'DewPointLowF', 'HumidityHighPercent', 'HumidityAvgPercent', 'HumidityLowPercent', 'SeaLevelPressureHighInches', 'SeaLevelPressureAvgInches', 'SeaLevelPressureLowInches', 'VisibilityHighMiles', 'VisibilityAvgMiles', 'VisibilityLowMiles', 'WindHighMPH', 'WindAvgMPH', 'WindGustMPH', 'Events']
Shape: (1319, 20)


#### Next we will perform linear interpolation of the missing data.

This means that we will use a linear algorithm to estimate the missing data. Linear interpolation assumes that there is a straight line between the points and the missing point will fall on that line. This is a good enough approximation for weather related data. Weather related data is typically a time series. Therefore, we do not want to drop rows from our data if possible. It is prefereable to estimate the missing values rather than remove the rows. However, if you have data from a single point in time, perhaps a better solution would be to remove the rows. 

If you would like to read more about linear interpolation, you can do so [here](https://en.wikipedia.org/wiki/Linear_interpolation).

In the following cell, use the `.interpolate()` function on the entire dataframe. This time pass the `inplace=False` argument to the function and assign the interpolated dataframe to a new variable called `austin_fixed` so that we can compare with `austin`.

In [42]:
# Your code here
# Perform linear interpolation on the remaining missing data
austin_fixed = weather_df.interpolate(inplace=False)


C:\Users\edumc\AppData\Local\Temp\ipykernel_55896\211276965.py:3: FutureWarning: DataFrame.interpolate with object dtype is deprecated and will raise in a future version. Call obj.infer_objects(copy=False) before interpolating instead.
  austin_fixed = weather_df.interpolate(inplace=False)


#### Check to make sure `austin_fixed` contains no missing data. Also check `austin` - it still contains missing data.

In [43]:
# Your code here
print("Missing values in austin_fixed:")
print(austin_fixed.isnull().sum().sum())  # Should be 0

# Also check austin still contains missing data
print(f"Missing values in original weather_df: {weather_df.isnull().sum().sum()}")


Missing values in austin_fixed:
0
Missing values in original weather_df: 80


# Challenge 3 - Processing the `Events` Column

#### Our dataframe contains one true text column - the Events column. We should evaluate this column to determine how to process it.

Use the `value_counts()` function to evaluate the contents of this column

In [50]:
# Your code here:
print("Events column value counts:")
print(austin_fixed['Events'].value_counts())


Events column value counts:
Events
                             903
Rain                         192
Rain , Thunderstorm          137
Fog , Rain , Thunderstorm     33
Fog                           21
Thunderstorm                  17
Fog , Rain                    14
Rain , Snow                    1
Fog , Thunderstorm             1
Name: count, dtype: int64


In [58]:
print(austin_fixed['Events'].unique())

['Rain , Thunderstorm' ' ' 'Rain' 'Fog' 'Rain , Snow' 'Fog , Rain'
 'Thunderstorm' 'Fog , Rain , Thunderstorm' 'Fog , Thunderstorm']


Reading the values of `Events` and reflecting what those values mean in the context of data, you realize this column indicates what weather events had happened in a particular day.

#### What is the largest number of events happened in a single day? Enter your answer in the next cell.

In [59]:
# Your answer:
# Look at the events to find the maximum number of events in a single day
# We need to count how many events are in each row by splitting on commas
event_counts = austin_fixed['Events'].apply(lambda x: len([event.strip() for event in str(x).split(',') if event.strip() and event.strip() != '']) if x and str(x).strip() else 0)

print(f"Maximum number of events in a single day: {event_counts.max()}")

# Let's also see which days had the most events
max_events_mask = event_counts == event_counts.max()
print(f"\nDays with {event_counts.max()} events:")
print(austin_fixed[max_events_mask][['Date', 'Events']])

Maximum number of events in a single day: 3

Days with 3 events:
           Date                     Events
114  2014-04-14  Fog , Rain , Thunderstorm
142  2014-05-12  Fog , Rain , Thunderstorm
157  2014-05-27  Fog , Rain , Thunderstorm
185  2014-06-24  Fog , Rain , Thunderstorm
186  2014-06-25  Fog , Rain , Thunderstorm
208  2014-07-17  Fog , Rain , Thunderstorm
257  2014-09-04  Fog , Rain , Thunderstorm
271  2014-09-18  Fog , Rain , Thunderstorm
336  2014-11-22  Fog , Rain , Thunderstorm
438  2015-03-04  Fog , Rain , Thunderstorm
483  2015-04-18  Fog , Rain , Thunderstorm
500  2015-05-05  Fog , Rain , Thunderstorm
508  2015-05-13  Fog , Rain , Thunderstorm
518  2015-05-23  Fog , Rain , Thunderstorm
520  2015-05-25  Fog , Rain , Thunderstorm
540  2015-06-14  Fog , Rain , Thunderstorm
542  2015-06-16  Fog , Rain , Thunderstorm
543  2015-06-17  Fog , Rain , Thunderstorm
545  2015-06-19  Fog , Rain , Thunderstorm
546  2015-06-20  Fog , Rain , Thunderstorm
547  2015-06-21  Fog , Rain , Th

#### We want to transform the string-type `Events` values to the numbers. This will allow us to apply machine learning algorithms easily.

How? We will create a new column for each type of events (i.e. *Rain*, *Snow*, *Fog*, *Thunderstorm*. In each column, we use `1` to indicate if the corresponding event happened in that day and use `0` otherwise.

Below we provide you a list of all event types. Loop the list and create a dummy column with `0` values for each event in `austin_fixed`. To create a new dummy column with `0` values, simply use `austin_fixed[event] = 0`.

In [60]:
event_list = ['Snow', 'Fog', 'Rain', 'Thunderstorm']

# Your code here
# Create dummy columns with 0 values for each event
for event in event_list:
    austin_fixed[event] = 0

# Print your new dataframe to check whether new columns have been created:
print("New columns added:")
print(austin_fixed[['Date', 'Events'] + event_list].head())
print(f"\nDataframe shape: {austin_fixed.shape}")


New columns added:
        Date               Events  Snow  Fog  Rain  Thunderstorm
0 2013-12-21  Rain , Thunderstorm     0    0     0             0
1 2013-12-22                          0    0     0             0
2 2013-12-23                          0    0     0             0
3 2013-12-24                          0    0     0             0
4 2013-12-25                          0    0     0             0

Dataframe shape: (1319, 24)


#### Next, populate the actual values in the dummy columns of  `austin_fixed`.

You will check the *Events* column. If its string value contains `Rain`, then the *Rain* column should be `1`. The same for `Snow`, `Fog`, and `Thunderstorm`.

*Hints:*

* Use [`pandas.Series.str.contains()`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.contains.html) to create the value series of each new column.

* What if the values you populated are booleans instead of numbers? You can cast the boolean values to numbers by using `.astype(int)`. For instance, `pd.Series([True, True, False]).astype(int)` will return a new series with values of `[1, 1, 0]`.

In [61]:
# Your code here
for event in event_list:
    # Use pandas.Series.str.contains() to create the value series
    contains_event = austin_fixed['Events'].str.contains(event, na=False)
    # Convert boolean series to integer (True becomes 1, False becomes 0)
    austin_fixed[event] = contains_event.astype(int)



#### Print out `austin_fixed` to check if the event columns are populated with the intended values

In [66]:
# Your code here
print("Sample of populated event columns:")
print(austin_fixed[['Date', 'Events']+ event_list].sample(10))

Sample of populated event columns:
           Date               Events  Snow  Fog  Rain  Thunderstorm
307  2014-10-24                          0    0     0             0
361  2014-12-17  Rain , Thunderstorm     0    0     1             1
96   2014-03-27                 Rain     0    0     1             0
285  2014-10-02  Rain , Thunderstorm     0    0     1             1
1191 2017-03-26                          0    0     0             0
1189 2017-03-24                 Rain     0    0     1             0
195  2014-07-04         Thunderstorm     0    0     0             1
838  2016-04-07                          0    0     0             0
1061 2016-11-16                          0    0     0             0
190  2014-06-29                          0    0     0             0


#### If your code worked correctly, now we can drop the `Events` column as we don't need it any more.

In [67]:
# Your code here
austin_fixed.drop('Events', axis=1, inplace=True)

In [68]:
# Print the dataframe to confirm Events column is removed
print("Columns after dropping Events:")
print(austin_fixed.columns.tolist())
print(f"Shape: {austin_fixed.shape}")

# Show a sample of the final data
print("\nSample of final data with event columns:")
print(austin_fixed[['Date'] + event_list].head())

Columns after dropping Events:
['Date', 'TempHighF', 'TempAvgF', 'TempLowF', 'DewPointHighF', 'DewPointAvgF', 'DewPointLowF', 'HumidityHighPercent', 'HumidityAvgPercent', 'HumidityLowPercent', 'SeaLevelPressureHighInches', 'SeaLevelPressureAvgInches', 'SeaLevelPressureLowInches', 'VisibilityHighMiles', 'VisibilityAvgMiles', 'VisibilityLowMiles', 'WindHighMPH', 'WindAvgMPH', 'WindGustMPH', 'Snow', 'Fog', 'Rain', 'Thunderstorm']
Shape: (1319, 23)

Sample of final data with event columns:
        Date  Snow  Fog  Rain  Thunderstorm
0 2013-12-21     0    0     1             1
1 2013-12-22     0    0     0             0
2 2013-12-23     0    0     0             0
3 2013-12-24     0    0     0             0
4 2013-12-25     0    0     0             0


In [69]:
# Final verification - let's make sure our encoding makes sense
print("Final event summary:")
total_event_days = austin_fixed[event_list].sum(axis=1).sum()
print(f"Total event occurrences across all days: {total_event_days}")

Final event summary:
Total event occurrences across all days: 635


In [70]:
# Check for days with multiple events
multiple_events = austin_fixed[austin_fixed[event_list].sum(axis=1) > 1]
print(f"Days with multiple events: {len(multiple_events)}")

if len(multiple_events) > 0:
    print("Sample days with multiple events:")
    print(multiple_events[['Date'] + event_list].head())

Days with multiple events: 186
Sample days with multiple events:
         Date  Snow  Fog  Rain  Thunderstorm
0  2013-12-21     0    0     1             1
33 2014-01-23     1    0     1             0
72 2014-03-03     0    0     1             1
77 2014-03-08     0    0     1             1
84 2014-03-15     0    1     1             0


# Challenge 4 - Processing The `Date` Column

The `Date` column is another non-numeric field in our dataset. A value in that field looks like `'2014-01-06'` which consists of the year, month, and day connected with hyphens. One way to convert the date string to numerical is using a similar approach as we used for `Events`, namely splitting the column into numerical `Year`, `Month`, and `Day` columns. In this challenge we'll show you another way which is to use the Python `datetime` library's `toordinal()` function. Depending on what actual machine learning analysis you will conduct, each approach has its pros and cons. Our goal today is to practice data preparation so we'll skip the discussion here.

Here you can find the [reference](https://docs.python.org/3/library/datetime.html) and [example](https://stackoverflow.com/questions/39846918/convert-date-to-ordinal-python) for `toordinal`. The basic process is to first convert the string to a `datetime` object using `datetime.datetime.strptime`, then convert the `datetime` object to numerical using `toordinal`.

#### In the cell below, convert the `Date` column values from string to numeric values using `toordinal()`.

In [71]:
# Your code here

import datetime

# First, let's examine the current Date column
print("Current Date column info:")
print(f"Data type: {austin_fixed['Date'].dtype}")
print("Sample values:")
print(austin_fixed['Date'].head())
print(f"Date range: {austin_fixed['Date'].min()} to {austin_fixed['Date'].max()}")

Current Date column info:
Data type: datetime64[ns]
Sample values:
0   2013-12-21
1   2013-12-22
2   2013-12-23
3   2013-12-24
4   2013-12-25
Name: Date, dtype: datetime64[ns]
Date range: 2013-12-21 00:00:00 to 2017-07-31 00:00:00


In [72]:
if austin_fixed['Date'].dtype == 'datetime64[ns]':
    print("Date is already datetime type, converting directly to ordinal...")
    austin_fixed['Date'] = austin_fixed['Date'].dt.date.apply(lambda x: x.toordinal())
else:
    # Method 2: If Date is string type, convert via strptime first
    print("Converting string dates to ordinal...")
    austin_fixed['Date'] = austin_fixed['Date'].apply(
        lambda x: datetime.datetime.strptime(str(x), '%Y-%m-%d').toordinal()
    )

Date is already datetime type, converting directly to ordinal...


#### Print `austin_fixed` to check your `Date` column.

In [73]:
austin_fixed.head(5)

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0  735223         74        60        45           67.0          49.0   
1  735224         56        48        39           43.0          36.0   
2  735225         58        45        32           31.0          27.0   
3  735226         61        46        31           36.0          28.0   
4  735227         58        50        41           44.0          40.0   

   DewPointLowF  HumidityHighPercent  HumidityAvgPercent  HumidityLowPercent  \
0          43.0                 93.0                75.0                57.0   
1          28.0                 93.0                68.0                43.0   
2          23.0                 76.0                52.0                27.0   
3          21.0                 89.0                56.0                22.0   
4          36.0                 86.0                71.0                56.0   

   ...  VisibilityHighMiles  VisibilityAvgMiles  VisibilityLowMiles  \
0  ...                 10.0                 7.0                 2.0   
1  ...                 10.0                10.0                 5.0   
2  ...                 10.0                10.0                10.0   
3  ...                 10.0                10.0                 7.0   
4  ...                 10.0                10.0                 7.0   

   WindHighMPH  WindAvgMPH  WindGustMPH  Snow  Fog  Rain  Thunderstorm  
0         20.0         4.0         31.0     0    0     1             1  
1         16.0         6.0         25.0     0    0     0             0  
2          8.0         3.0         12.0     0    0     0             0  
3         12.0         4.0         20.0     0    0     0             0  
4         10.0         2.0         16.0     0    0     0             0  

[5 rows x 23 columns]

# Challenge 5 - Sampling and Holdout Sets

#### Now that we have processed the data for machine learning, we will separate the data to test and training sets.

We first train the model using only the training set. We check our metrics on the training set. We then apply the model to the test set and check our metrics on the test set as well. If the metrics are significantly more optimal on the training set, then we know we have overfit our model. We will need to revise our model to ensure it will be more applicable to data outside the test set.

#### In the next cells we will separate the data into a training set and a test set using the `train_test_split()` function in scikit-learn.

When using `scikit-learn` for machine learning, we first separate the data to predictor and response variables. This is the standard way of passing datasets into a model in `scikit-learn`. The `scikit-learn` will then find out whether the predictors and responses fit the model.

In the next cell, assign the `TempAvgF` column to `y` and the remaining columns to `X`. Your `X` should be a subset of `austin_fixed` containing the following columns: 

```['Date',
 'TempHighF',
 'TempLowF',
 'DewPointHighF',
 'DewPointAvgF',
 'DewPointLowF',
 'HumidityHighPercent',
 'HumidityAvgPercent',
 'HumidityLowPercent',
 'SeaLevelPressureHighInches',
 'SeaLevelPressureAvgInches',
 'SeaLevelPressureLowInches',
 'VisibilityHighMiles',
 'VisibilityAvgMiles',
 'VisibilityLowMiles',
 'WindHighMPH',
 'WindAvgMPH',
 'WindGustMPH',
 'Snow',
 'Fog',
 'Rain',
 'Thunderstorm']```
 
 Your `y` should be a subset of `austin_fixed` containing one column `TempAvgF`.

In [74]:
# Your code here:
# Assign the TempAvgF column to y (target variable)
y = austin_fixed['TempAvgF']

# Assign the remaining columns to X (predictor variables)
# All columns except TempAvgF
X = austin_fixed.drop('TempAvgF', axis=1)

print("Target variable (y):")
print(f"Shape: {y.shape}")
print(f"Variable: TempAvgF")
print("Sample values:")
print(y.head())

print(f"\nPredictor variables (X):")
print(f"Shape: {X.shape}")
print("Columns:")
print(X.columns.tolist())

Target variable (y):
Shape: (1319,)
Variable: TempAvgF
Sample values:
0    60
1    48
2    45
3    46
4    50
Name: TempAvgF, dtype: int64

Predictor variables (X):
Shape: (1319, 22)
Columns:
['Date', 'TempHighF', 'TempLowF', 'DewPointHighF', 'DewPointAvgF', 'DewPointLowF', 'HumidityHighPercent', 'HumidityAvgPercent', 'HumidityLowPercent', 'SeaLevelPressureHighInches', 'SeaLevelPressureAvgInches', 'SeaLevelPressureLowInches', 'VisibilityHighMiles', 'VisibilityAvgMiles', 'VisibilityLowMiles', 'WindHighMPH', 'WindAvgMPH', 'WindGustMPH', 'Snow', 'Fog', 'Rain', 'Thunderstorm']


In [76]:
# Let's verify X contains the expected columns from the challenge
expected_columns = [
 'Date',
 'TempHighF',
 'TempLowF',
 'DewPointHighF',
 'DewPointAvgF',
 'DewPointLowF',
 'HumidityHighPercent',
 'HumidityAvgPercent',
 'HumidityLowPercent',
 'SeaLevelPressureHighInches',
 'SeaLevelPressureAvgInches',
 'SeaLevelPressureLowInches',
 'VisibilityHighMiles',
 'VisibilityAvgMiles',
 'VisibilityLowMiles',
 'WindHighMPH',
 'WindAvgMPH',
 'WindGustMPH',
 'Snow',
 'Fog',
 'Rain',
 'Thunderstorm']

print("Expected columns vs Actual columns:")
print(f"Expected: {len(expected_columns)} columns")
print(f"Actual: {len(X.columns)} columns")

# Check if we have all expected columns
missing_cols = set(expected_columns) - set(X.columns)
extra_cols = set(X.columns) - set(expected_columns)

if missing_cols:
    print(f"Missing columns: {missing_cols}")
if extra_cols:
    print(f"Extra columns: {extra_cols}")
if not missing_cols and not extra_cols:
    print("✓ All columns match perfectly!")

Expected columns vs Actual columns:
Expected: 22 columns
Actual: 22 columns
✓ All columns match perfectly!


In the next cell, import `train_test_split` from `sklearn.model_selection`

In [77]:
#Your code here:
# Import train_test_split from sklearn.model_selection
from sklearn.model_selection import train_test_split

print("train_test_split imported successfully!")

train_test_split imported successfully!


Now that we have split the data to predictor and response variables and imported the `train_test_split()` function, split `X` and `y` into `X_train`, `X_test`, `y_train`, and `y_test`. 80% of the data should be in the training set and 20% in the test set. `train_test_split()` reference can be accessed [here](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html).


Enter your code in the cell below:

In [79]:
#Your code here:
# Split X and y into training and test sets
# 80% training, 20% test (test_size=0.2)
# random_state for reproducibility
X_train, X_test, y_train, y_test = train_test_split(X, y,test_size=0.2, random_state=42)

print("Data split completed!")
print(f"Training set size: {X_train.shape[0]} samples ({X_train.shape[0]/len(X)*100:.1f}%)")
print(f"Test set size: {X_test.shape[0]} samples ({X_test.shape[0]/len(X)*100:.1f}%)")
print(f"Total samples: {len(X)}")

print(f"\nX_train shape: {X_train.shape}")
print(f"X_test shape: {X_test.shape}")
print(f"y_train shape: {y_train.shape}")
print(f"y_test shape: {y_test.shape}")

Data split completed!
Training set size: 1055 samples (80.0%)
Test set size: 264 samples (20.0%)
Total samples: 1319

X_train shape: (1055, 22)
X_test shape: (264, 22)
y_train shape: (1055,)
y_test shape: (264,)


In [80]:
# Verify the split worked correctly
print("Verification of train/test split:")

# Check that we didn't lose any data
total_train_test = len(X_train) + len(X_test)
print(f"Original data: {len(X)} samples")
print(f"Train + Test: {total_train_test} samples")
print(f"Data preserved: {total_train_test == len(X)}")

# Check target variable distribution
print(f"\nTarget variable (TempAvgF) statistics:")
print("Training set:")
print(y_train.describe())
print("\nTest set:")
print(y_test.describe())
print("\nOriginal set:")
print(y.describe())

# Show sample of training data
print(f"\nSample of training data (X_train):")
print(X_train.head())
print(f"\nCorresponding target values (y_train):")
print(y_train.head())

Verification of train/test split:
Original data: 1319 samples
Train + Test: 1319 samples
Data preserved: True

Target variable (TempAvgF) statistics:
Training set:
count    1055.000000
mean       70.628436
std        13.966018
min        29.000000
25%        62.000000
50%        73.000000
75%        83.000000
max        93.000000
Name: TempAvgF, dtype: float64

Test set:
count    264.000000
mean      70.700758
std       14.387542
min       29.000000
25%       61.500000
50%       73.000000
75%       82.250000
max       92.000000
Name: TempAvgF, dtype: float64

Original set:
count    1319.000000
mean       70.642911
std        14.045904
min        29.000000
25%        62.000000
50%        73.000000
75%        83.000000
max        93.000000
Name: TempAvgF, dtype: float64

Sample of training data (X_train):
        Date  TempHighF  TempLowF  DewPointHighF  DewPointAvgF  DewPointLowF  \
598   735821        105        75           70.0         69.25          67.5   
1213  736436         80  

In [81]:
# Additional checks to ensure everything is correct
print("Additional verification:")

# Check for any missing values in the splits
print(f"Missing values in X_train: {X_train.isnull().sum().sum()}")
print(f"Missing values in X_test: {X_test.isnull().sum().sum()}")
print(f"Missing values in y_train: {y_train.isnull().sum()}")
print(f"Missing values in y_test: {y_test.isnull().sum()}")

# Verify the indices are different (no overlap)
train_indices = set(X_train.index)
test_indices = set(X_test.index)
overlap = train_indices.intersection(test_indices)
print(f"Index overlap between train and test: {len(overlap)} (should be 0)")

# Check the date range in both sets
if 'Date' in X_train.columns:
    print(f"\nDate range in training set: {X_train['Date'].min()} to {X_train['Date'].max()}")
    print(f"Date range in test set: {X_test['Date'].min()} to {X_test['Date'].max()}")

Additional verification:
Missing values in X_train: 0
Missing values in X_test: 0
Missing values in y_train: 0
Missing values in y_test: 0
Index overlap between train and test: 0 (should be 0)

Date range in training set: 735223 to 736541
Date range in test set: 735246 to 736536


#### Congratulations! Now you have finished the preparation of the dataset!

# Bonus Challenge 1

#### While the above is the common practice to prepare most datasets, when it comes to time series data, we sometimes do not want to randomly select rows from our dataset.

This is because many time series algorithms rely on observations having equal time distances between them. In such cases, we typically select the majority of rows as the test data and the last few rows as the training data. We don't use `train_test_split()` to select the train/test data because it returns random selections.

In the following cell, compute the number of rows that account for 80% of our data and round it to the next integer. Assign this number to `ts_rows`.

In [82]:
# Your code here:
import math

# Compute the number of rows that account for 80% of our data
total_rows = len(austin_fixed)
ts_rows = math.ceil(total_rows * 0.8)  # Round up to next integer

print(f"Total rows in dataset: {total_rows}")
print(f"80% of data: {total_rows * 0.8}")
print(f"Rounded up (ts_rows): {ts_rows}")
print(f"Training set will be: {ts_rows} rows ({ts_rows/total_rows*100:.1f}%)")
print(f"Test set will be: {total_rows - ts_rows} rows ({(total_rows - ts_rows)/total_rows*100:.1f}%)")

Total rows in dataset: 1319
80% of data: 1055.2
Rounded up (ts_rows): 1056
Training set will be: 1056 rows (80.1%)
Test set will be: 263 rows (19.9%)


Assign the first `ts_rows` rows of `X` to `X_ts_train` and the remaining rows to `X_ts_test`.

In [83]:
# Your code here:
# Assign the first ts_rows rows of X to X_ts_train and the remaining rows to X_ts_test
X_ts_train = X.iloc[:ts_rows]  # First ts_rows rows
X_ts_test = X.iloc[ts_rows:]   # Remaining rows

print("Time series split for X (features):")
print(f"X_ts_train shape: {X_ts_train.shape}")
print(f"X_ts_test shape: {X_ts_test.shape}")

# Verify we didn't lose any data
print(f"Total samples: {len(X_ts_train) + len(X_ts_test)} (should equal {total_rows})")
print(f"No data loss: {len(X_ts_train) + len(X_ts_test) == total_rows}")

Time series split for X (features):
X_ts_train shape: (1056, 22)
X_ts_test shape: (263, 22)
Total samples: 1319 (should equal 1319)
No data loss: True


Assign the first `ts_rows` rows of `y` to `y_ts_train` and the remaining rows to `y_ts_test`.

In [84]:
# Your code here:
# Assign the first ts_rows rows of y to y_ts_train and the remaining rows to y_ts_test
y_ts_train = y.iloc[:ts_rows]  # First ts_rows rows
y_ts_test = y.iloc[ts_rows:]   # Remaining rows

print("Time series split for y (target):")
print(f"y_ts_train shape: {y_ts_train.shape}")
print(f"y_ts_test shape: {y_ts_test.shape}")

# Verify we didn't lose any data
print(f"Total samples: {len(y_ts_train) + len(y_ts_test)} (should equal {total_rows})")
print(f"No data loss: {len(y_ts_train) + len(y_ts_test) == total_rows}")

Time series split for y (target):
y_ts_train shape: (1056,)
y_ts_test shape: (263,)
Total samples: 1319 (should equal 1319)
No data loss: True


In [85]:
# Let's verify our time series split makes chronological sense
print("Verification of time series split:")

# Check the date ranges
if 'Date' in X.columns:
    print(f"Training set date range:")
    print(f"  From: {datetime.date.fromordinal(int(X_ts_train['Date'].min()))}")
    print(f"  To:   {datetime.date.fromordinal(int(X_ts_train['Date'].max()))}")
    
    print(f"Test set date range:")
    print(f"  From: {datetime.date.fromordinal(int(X_ts_test['Date'].min()))}")
    print(f"  To:   {datetime.date.fromordinal(int(X_ts_test['Date'].max()))}")
    
    # Verify chronological order
    train_max_date = X_ts_train['Date'].max()
    test_min_date = X_ts_test['Date'].min()
    print(f"\nChronological order check:")
    print(f"Latest training date: {datetime.date.fromordinal(int(train_max_date))}")
    print(f"Earliest test date: {datetime.date.fromordinal(int(test_min_date))}")
    print(f"Proper time series order: {train_max_date < test_min_date}")

Verification of time series split:
Training set date range:
  From: 2013-12-21
  To:   2016-11-10
Test set date range:
  From: 2016-11-11
  To:   2017-07-31

Chronological order check:
Latest training date: 2016-11-10
Earliest test date: 2016-11-11
Proper time series order: True


In [86]:
# Compare the two splitting approaches
print("="*60)
print("COMPARISON: Random Split vs Time Series Split")
print("="*60)

print("RANDOM SPLIT (from Challenge 5):")
print(f"X_train shape: {X_train.shape}")
print(f"X_test shape: {X_test.shape}")
if 'Date' in X_train.columns:
    print(f"Train date range: {datetime.date.fromordinal(int(X_train['Date'].min()))} to {datetime.date.fromordinal(int(X_train['Date'].max()))}")
    print(f"Test date range: {datetime.date.fromordinal(int(X_test['Date'].min()))} to {datetime.date.fromordinal(int(X_test['Date'].max()))}")

print(f"\nTIME SERIES SPLIT (Bonus Challenge):")
print(f"X_ts_train shape: {X_ts_train.shape}")
print(f"X_ts_test shape: {X_ts_test.shape}")
if 'Date' in X_ts_train.columns:
    print(f"Train date range: {datetime.date.fromordinal(int(X_ts_train['Date'].min()))} to {datetime.date.fromordinal(int(X_ts_train['Date'].max()))}")
    print(f"Test date range: {datetime.date.fromordinal(int(X_ts_test['Date'].min()))} to {datetime.date.fromordinal(int(X_ts_test['Date'].max()))}")

COMPARISON: Random Split vs Time Series Split
RANDOM SPLIT (from Challenge 5):
X_train shape: (1055, 22)
X_test shape: (264, 22)
Train date range: 2013-12-21 to 2017-07-31
Test date range: 2014-01-13 to 2017-07-26

TIME SERIES SPLIT (Bonus Challenge):
X_ts_train shape: (1056, 22)
X_ts_test shape: (263, 22)
Train date range: 2013-12-21 to 2016-11-10
Test date range: 2016-11-11 to 2017-07-31


In [87]:
# Show samples from each time series set
print("\nSample from TIME SERIES TRAINING set (earliest data):")
print(X_ts_train[['Date', 'TempHighF', 'TempLowF']].head())

print("\nSample from TIME SERIES TEST set (latest data):")
print(X_ts_test[['Date', 'TempHighF', 'TempLowF']].head())

# Show target variable statistics for time series split
print(f"\nTarget variable statistics (Time Series Split):")
print("Training set (y_ts_train):")
print(y_ts_train.describe())
print("\nTest set (y_ts_test):")
print(y_ts_test.describe())


Sample from TIME SERIES TRAINING set (earliest data):
     Date  TempHighF  TempLowF
0  735223         74        45
1  735224         56        39
2  735225         58        32
3  735226         61        31
4  735227         58        41

Sample from TIME SERIES TEST set (latest data):
        Date  TempHighF  TempLowF
1056  736279         74        55
1057  736280         77        57
1058  736281         76        54
1059  736282         82        56
1060  736283         86        59

Target variable statistics (Time Series Split):
Training set (y_ts_train):
count    1056.000000
mean       70.697917
std        14.248527
min        29.000000
25%        61.000000
50%        73.000000
75%        83.000000
max        92.000000
Name: TempAvgF, dtype: float64

Test set (y_ts_test):
count    263.000000
mean      70.422053
std       13.224739
min       29.000000
25%       62.000000
50%       71.000000
75%       80.000000
max       93.000000
Name: TempAvgF, dtype: float64


In [88]:
# Final verification
print("="*60)
print("BONUS CHALLENGE 1 COMPLETED!")
print("="*60)

# Verify no index overlap
ts_train_indices = set(X_ts_train.index)
ts_test_indices = set(X_ts_test.index)
ts_overlap = ts_train_indices.intersection(ts_test_indices)

print(f"✓ Time series training set: {len(X_ts_train)} samples")
print(f"✓ Time series test set: {len(X_ts_test)} samples")
print(f"✓ Total samples preserved: {len(X_ts_train) + len(X_ts_test)} = {total_rows}")
print(f"✓ No index overlap: {len(ts_overlap)} overlapping indices")
print(f"✓ Chronological order maintained")
print(f"✓ Training on historical data, testing on recent data")

# Show the exact split point
split_date = datetime.date.fromordinal(int(X_ts_train['Date'].max()))
print(f"✓ Split point: {split_date}")

BONUS CHALLENGE 1 COMPLETED!
✓ Time series training set: 1056 samples
✓ Time series test set: 263 samples
✓ Total samples preserved: 1319 = 1319
✓ No index overlap: 0 overlapping indices
✓ Chronological order maintained
✓ Training on historical data, testing on recent data
✓ Split point: 2016-11-10
